pkg/controllers/disruption/controller.go

```diff
@@ -179,194 +161,60 @@ func (c *Controller) disrupt(ctx context.Context, disruption Method) (bool, erro
 
-	reason := fmt.Sprintf("%s/%s", m.Type(), cmd.Action())
-	if cmd.Action() == ReplaceAction {
-		if err := c.launchReplacementNodeClaims(ctx, m, cmd); err != nil {
+	stateNodes := lo.Map(cmd.candidates, func(c *Candidate, _ int) *state.StateNode {
+		return c.StateNode
+	})
+	// Cordon the old nodes before we launch the replacements to prevent new pods from scheduling to the old nodes
+	if err := state.RequireNoScheduleTaint(ctx, c.kubeClient, true, stateNodes...); err != nil {
+		return multierr.Append(fmt.Errorf("cordoning nodes, %w", err), state.RequireNoScheduleTaint(ctx, c.kubeClient, false, stateNodes...))
+	}
+	var nodeClaimKeys []nodeclaim.Key
+	var err error
+	if len(cmd.replacements) > 0 {
+		if nodeClaimKeys, err = c.launchReplacementNodeClaims(ctx, m, cmd); err != nil {
 			// If we failed to launch the replacement, don't disrupt.  If this is some permanent failure,
-			// we don't want to disrupt workloads with no way to provision new NodeClaims for them.
-			return fmt.Errorf("launching replacement, %w", err)
+			// we don't want to disrupt workloads with no way to provision new nodes for them.
+			return multierr.Append(fmt.Errorf("launching replacement nodeclaim, %w", err), state.RequireNoScheduleTaint(ctx, c.kubeClient, false, stateNodes...))
 		}
 	}
+	providerIDs := lo.Map(cmd.candidates, func(c *Candidate, _ int) string { return c.ProviderID() })
+	// We have the new NodeClaims created at the API server so mark the old NodeClaims for deletion
+	c.cluster.MarkForDeletion(providerIDs...)
 
-	for _, candidate := range cmd.candidates {
-		c.recorder.Publish(disruptionevents.Terminating(candidate.Node, candidate.NodeClaim, reason)...)
-
-		if err := nodeclaimutil.Delete(ctx, c.kubeClient, candidate.NodeClaim); err != nil {
-			if !errors.IsNotFound(err) {
-				logging.FromContext(ctx).Errorf("terminating, %s", err)
-			}
-			continue
-		}
-		nodeclaimutil.TerminatedCounter(candidate.NodeClaim, reason).Inc()
-	}
-
-	// We wait for NodeClaims to delete to ensure we don't start another round of disruption
-	// until this node is fully deleted.
-	for _, oldCandidate := range cmd.candidates {
-		c.waitForDeletion(ctx, oldCandidate.NodeClaim)
+	if err := c.Queue.Add(orchestration.NewCommand(nodeClaimKeys,
+		lo.Map(cmd.candidates, func(c *Candidate, _ int) *state.StateNode { return c.StateNode }), c.clock.Now(), m.Type(), m.ConsolidationType())); err != nil {
+		c.cluster.UnmarkForDeletion(providerIDs...)
+		return fmt.Errorf("adding command to queue, %w", err)
 	}
 	return nil
 }
 
-// launchReplacementNodeClaims launches replacement NodeClaims and blocks until it is ready
+// launchReplacementNodeClaims launches replacement NodeClaims
 // nolint:gocyclo
-func (c *Controller) launchReplacementNodeClaims(ctx context.Context, m Method, cmd Command) error {
+func (c *Controller) launchReplacementNodeClaims(ctx context.Context, m Method, cmd Command) ([]nodeclaim.Key, error) {
```

Comment:
This might be a misnomer since we aren't launching the NodeClaims here

Reply:
Ah fair distinction. It should be `createNodeClaims`